Implement Taiko Difficulty Calculator

hongaaronc opened the pull request:
## Summary
This PR implements the Taiko Difficulty Calculator for SR calculation.

This is pretty useful, since recently [a Taiko MV plugin](https://github.com/Hiviexd/MVTaikoChecks) has now been released and frequently used by mappers/modders. And since SR calc is not implemented, all star ratings are 0.00*, which makes any diff with a custom name default to an "Easy" (Kantan in taiko). (reference: https://github.com/Hiviexd/MVTaikoChecks/issues/6)

As far as approach taken, I aimed to port [the osu!lazer diff calc](https://github.com/ppy/osu/blob/master/osu.Game.Rulesets.Taiko/Difficulty/TaikoDifficultyCalculator.cs) while making as few changes to the original source as possible. I also tried to keep the existing `MapsetParser` code as intact as possible.

However, I will discuss the parts I did have to change.

## MapsetParser Classes which differ from their original implementation
* `Skill` was originally basically a superclass which handled strain/decay. In Lazer implementation, `Skill` is much more abstract and generic, and there are separate `StrainSkill` and `StrainDecaySkill` subclasses instead.
  * I basically changed this to be more like the lazer implementation, trying to keep it as close to lazer source as possible
  * `Skill` also previously tracked the "previous" hit object from the current one being calculated. In lazer, this is now tracked within `DifficultyHitObject` so I removed this from `Skill`
  * Other classes have been modified slightly to accomodate these changes, including `Aim`, `Speed`, and `DifficultyCalculator`
* `DifficultyHitObject` now has members/methods to track Next and Previous objects, Index, and StartTime and EndTimes
* `Beatmap` is modified to use the `TaikoDifficultyCalculator` when taiko mode is detected
  * The parsing was also changed if the game mode is taiko, so that it will use the new `Hit` class instead of `Circle`
  
 ## Lazer classes which differ from their lazer implementation
 * All classes using `clockrate` or mods have had that removed, since it's not really relevant for MV purposes
 * Any stuff with serialization was also removed
 * `Hit` is quite different from the lazer implementation. In my implementation it's basically similar to a `Circle` but it will figure out if its a rim/centre hit automatically from the args in the `.osu` file
 * Pretty much everything else is almost verbatim the same as the lazer implementation
 
 ## Things that could be maybe changed / need 2nd thoughts in review
 * The `Beatmap` parsing changes look kinda hacky to me atm, however doing this allows us to keep the classes ported from lazer almost completely 1:1 verbatim
   * I wouldn't be that opposed to removing the `Hit` and `HitType` and all associated classes entirely, but it would mean the other ported classes would resemble their original lazer implementations less closely (is this worth?)
 * Porting the `IBeatmapDifficultyInfo` class is probably unnecessary since I used like 1 method from it. Not sure which direction to go from here tho - should I just reduce that class to the 1 method or maybe extract that method to somewhere else?

## Testing
* I made a check in MV plugin that just emits the star rating for debugging (can be downloaded [here](https://nostril.s-ul.eu/en8Zt6tT) if you want)
* Standard SR still works (obviously based on old algorithm, but it's good enough)
* Taiko SR looks pretty accurate - sometimes it's off by like +/- 0.01 SR or so though, I'm not sure what the discrepency is but seems very trivial

Naxesss commented:
Did a read-over and some testing and it seems to work well. There's one issue I wasn't able to fix (more on that later), and a few things I figured I'd change with regards to how it was coded, notably:
1. `Hit` and `HitType` seem a bit superfluous, and is giving a new name to what is really just a `Circle`. It's the same kind of object, we just want to ask it taiko-specific questions in taiko contexts. So I've added `hitobjects.taiko.TaikoExtensions` (requiring a "/.../.taiko" import), which extends hit objects with taiko methods, like `IsDon` and `IsFinisher`. This is basically what the taiko plugin does, and seems a bit easier to work with I think.
2. `IBeatmapDifficultyInfo` and `HitWindows` had some overlapping functionality with `DifficultySettings`. I've removed `IBeatmapDifficultyInfo` and let `DifficultySettings` take its use case. I've also removed hit-window-related code from `DifficultySettings` in favor of `HitWindows` and lazer's `OsuHitWindows`.
3. I set up some foundation code so skill charts in the overview section of **MapsetVerifierBackend** support these changes, including initializing `Skills` in `TaikoDifficultyAttributes` and [now showing the (approximate) star rating for taiko](https://github.com/Naxesss/MapsetParser/assets/30292137/91ed1588-22d3-4985-85fc-438e1bffead5). For reference, this is how skill charts look in standard:

![image](https://github.com/Naxesss/MapsetParser/assets/30292137/af5eaf4b-f74b-41c2-818d-de789d6f0b8c)

If any of those changes don't make sense, do let me know.

---

The remaining issue is that `Peaks` is just straight up extending `Skill`, so it has no strain peak data, unlike `Rhythm`, `Stamina`, and `Colour` which it combines. This means the skill charts in the overview section don't show how the SR fluctuates throughout taiko maps. As far as I tried fixing it (using the three skills in place of `Peaks`), the SR values kinda exploded, and not just by some linear amount I'd be able to divide away. [Right now it looks pretty empty.](https://github.com/Naxesss/MapsetParser/assets/30292137/a33ab8f4-8d4a-4127-8f5b-26b321362aff) I could probably get rid of that section for taiko if we end up not fixing it. It would be cool to have, though.

What do you think, is it worth trying to fix?

---

> Maybe could change the SR -> Difficulty ranges for taiko? Like make this a mapping per game mode?

[This page seems to imply that the ranges are gamemode-agnostic](https://osu.ppy.sh/wiki/en/Beatmap/Difficulty), is that not the case? Am not very familiar with how difficulty cut-offs work in other gamemodes, always kinda assumed SR was scaled to make it work.

Hiviexd commented:
imo the difficulty cutoffs should stay gamemode-agnostic because of simplicity + lack of agreed-upon cutoffs for taiko, anyways mv prioritizes diffnaming over sr when checking, so a 3* futsuu would still be treated as a futsuu

hongaaronc commented:
Nice! Good stuff Naxess.... I think your changes address everything that I was feeling iffy on for code quality. It looks good!

Only made one small tweak: the `TaikoExtensions.isDon(..)` was checking for the opposite (whistles / claps in taiko actually mean the note is a kat), so I just flipped it so should be good now. I think this doesn't actually make a difference for diff calc, since I think the calc only cares about relative color changes but good to fix anyway since this seems like a nice utility method.

> What do you think, is it worth trying to fix?

Wow I didn't even realize these graphs were a thing, since I was so used to seeing this part being blank (taiko mapper moment). These look really awesome and helpful, so yeah I think it would be worth trying to fix. I'll take a look into it.

> [This page seems to imply that the ranges are gamemode-agnostic](https://osu.ppy.sh/wiki/en/Beatmap/Difficulty), is that not the case? Am not very familiar with how difficulty cut-offs work in other gamemodes, always kinda assumed SR was scaled to make it work.

Actually I'm fine with what hivie said, so fine to keep as-is.

hongaaronc commented:
Added in my changes to change `Peaks` into a `StrainSkill`.

I'm now able to see some stuff graphed which seems very accurate to where the actual diffspikes are in this beatmap. However, the scale seems a bit weird - but I mean so is the osu! standard one so maybe it's fine... I did use the same transformation function as used in `TaikoDifficultyCalculator` so I suppose that's just how it is.
![image](https://github.com/Naxesss/MapsetParser/assets/5133530/3a410a9f-6ef9-435e-af22-da25e75e1795)

Depends on:
* https://github.com/Naxesss/MapsetVerifierBackend/pull/1
* https://github.com/Naxesss/MapsetVerifierBackend/pull/2

hongaaronc commented:
Added another thing along with these changes - a scroll speed graph for Taiko. I have implemented this as an additional skill, but also added a new flag - `useInStarRating` - which allows us to ignore this new skill in the accumulated star rating graph.

Some modders already use a similar tool to this: https://github.com/Iojioji/Taiko-SV-Viewer . This basically does the same thing but in MV.

As far as usecases go, it can be used to check:
* If a loweer diff has more/more extreme SV than a higher diff in a spread
* If there are any unintended SV anomalies
* For a complex timing map, if SV equalization was done properly

Regarding the newly added `useInStarRating` flag, I've also included a corresponding supporting change in MapsetVerifierBackend: https://github.com/Naxesss/MapsetVerifierBackend/pull/2/commits/04f1eb82ae5afbf9ecfed00e41666f54e9da2c96
